[Clang] Implement C++26’s P2893R3 ‘Variadic friends’

Sirraide opened the pull request:
Implement P2893R3 ‘Variadic friends’ for C++26.

The implementation proper is mostly done, from what I can tell, but I have a few questions, so I thought I’d be best to list them all here (some are for me to investigate, some I’d appreciate it if anyone would happen to know the answer to them):
- [x] ~~I don’t know much about the `ASTImporter`, so I’m not entirely sure if I did everything right in there.~~ No longer relevant.
- [x] Still need to add tests for [CWG 2917](https://cplusplus.github.io/CWG/issues/2917.html), which I filed while working on this, to the DR test suite proper. (@Endilll Sorry for asking this *again*, but do I need to run the script that updates the DR tests as part of this, or should that be done in a separate pr before this?)
- [x] Do we want to expose this in earlier language modes? The syntax wasn’t valid before, so I *believe* there shouldn’t be an issue with that, but I’m not sure.
- [x] There are some edge cases that mostly involve what we treat as unsupported friend declarations that I still need to either reject or also convert to unsupported friend declarations.
- [x] ~~Is there ever a situation that would cause us to have to instantiate a `FriendPackDecl` in the `TemplateDeclInstantiator`? We do that for `UsingPackDecl`s, but those are sufficiently different enough from the former that I’m not sure about it. I at least haven’t been able to come up w/ any cases that would cause us to try and instantiate one. (We’ll revisit this later)~~ No longer relevant.
- [x] The diagnostics for some invalid declarations (e.g. CWG 2917) could still be better imo.
- [x] I had to add a hack to the decl printer to get e.g. `friend int, long;` to be printed like that instead of as separate declarations (which they currently are in the AST). Is that fine or is that too much of a hack? Update: Too much of a hack. Removed.
- [x] Whether an ellipsis is present after a friend-type-specifier is currently handled via a separate parameter in some of the Sema functions. It would be possible to instead store that information in the `DeclSpec`, but I haven’t written too much parser code in Clang before, and wasn’t sure how strict we are wrt putting things in that class because I don’t think the `...` is part of the specifier according to the grammar. Update: Seems like it’s fine as-is.

Thanks also to @zyn0217, who was working on this before me; I’ve taken over working on this since they didn’t have time for it anymore. I’ve incorporated parts of their WIP patch, so I’ll credit them as a co-author whenever this gets merged.

This closes #98587.

Sirraide committed:
[Parser] Parse variadic friends

Sirraide committed:
[Sema] Instantiate variadic friends

Sirraide committed:
[ASTPrinter] Group friend decls

Sirraide committed:
[Tests] Move some tests

Sirraide committed:
Feature-test macro + release notes

Sirraide committed:
clang-format

zyn0217 commented on clang/lib/AST/ASTImporter.cpp:
```diff
@@ -4437,6 +4441,31 @@ ExpectedDecl ASTNodeImporter::VisitFriendDecl(FriendDecl *D) {
+  if (Error Err = ImportDeclContext(D, DC, LexicalDC))
+    return std::move(Err);
+
+  auto ToInstantiatedFromFriendOrErr =
+      Importer.Import(D->getInstantiatedFromFriendDecl());
```
nit: It would be great to spell out the type name if it is not too long - I'm also unfamiliar with the ASTImporter, though :)

Sirraide replied:
Hmm, I basically just copy-pasted stuff from the functions that handle `UsingPackDecl`s and `FriendDecl`s, and it seems that those always use `auto`, so this would probably be better left for a separate NFC patch imo, but I suppose I could change it in this function here.

zyn0217 commented on clang/lib/AST/DeclPrinter.cpp:
```diff
@@ -487,7 +487,26 @@ void DeclPrinter::VisitDeclContext(DeclContext *DC, bool Indent) {
+      // now before this loop if there are no subsequent friends) will be
+      // skipped by the '++D' of the outer loop.
+      for (DeclContext::decl_iterator It; It = std::next(D), It != DEnd; ++D) {
+        auto NextFriend = dyn_cast<FriendDecl>(*It);
```
ditto, we prefer `auto *`

zyn0217 commented on clang/lib/Frontend/InitPreprocessor.cpp:
```diff
@@ -754,6 +754,10 @@ static void InitializeCPlusPlusFeatureTestMacros(const LangOptions &LangOpts,
     Builder.defineMacro("__cpp_multidimensional_subscript", "202211L");
     Builder.defineMacro("__cpp_auto_cast", "202110L");
   }
+  // C++26 features.
+  if (LangOpts.CPlusPlus26) {
+    Builder.defineMacro("__cpp_variadic_friend", "202403L");
+  }
```
I thought we already had an umbrella for C++26 features e.g. pack indexing expressions... @cor3ntin is it an oversight?

Sirraide replied:
We actually don’t define `__cpp_pack_indexing` yet, and the other C++26 features that we support that have a feature-test macro are also exposed in previous language modes, so this would be the first one that’s (currently) C++26 only (though, again, maybe we want to expose this one in older language modes too?).

I don’t actually know if this is the case, but I’m assuming the reason why we don’t define `__cpp_pack_indexing` is because there is a chance that the syntax might change or something like that?

Sirraide replied:
Update: `__cpp_pack_indexing` has since been added in a separate pr, but since we’re splitting this into multiple patches now, we won’t be adding a feature test macro for this just yet.

zyn0217 reviewed (commented):
Thanks so much for continuing the effort to implement this! I haven't dived into the details yet, so here are some nits from me.

(I'd also prefer the pronouns he/him if you'd like. ;)

Sirraide commented:
CI failure seems unrelated

cor3ntin commented on clang/lib/Sema/SemaTemplateInstantiateDecl.cpp:
```diff
@@ -1442,8 +1442,49 @@ Decl *TemplateDeclInstantiator::VisitFriendDecl(FriendDecl *D) {
+                FriendDecl::Create(SemaRef.Context, Owner, D->getLocation(),
+                                   TSI, D->getFriendLoc());
+
+            FD->setAccess(AS_public);
+            Owner->addDecl(FD);
```
Should we call `setObjectOfFriendDecl` here?

Sirraide replied:
We don’t do that when we instantiate `FriendDecl`s below, which is why I didn’t add it here.

cor3ntin replied:
I think that's probably a bug? @AaronBallman

AaronBallman replied:
I *think* that is not a bug (I think this gets called on the thing that is being friended, not on the friend itself), but I'd appreciate a second opinion from @zygoloid or @rjmccall

cor3ntin reviewed (commented):
Looks generally good from a cursory review.
I think we want to add tests for modules / in the test/PCH directory.

We probably want that as an extension indeed, no reason not to

We probably want a separate PR to add the pack indexing macro. I did not add it initially because i was concerned the implementation was not mature enough, but i think we fond most of the bugs.

I think it's fine to set the feature macro for that now with the assumption that we will find all the bugs in the next 6 months.

There seem to be some missing components, like the json node printer.

Sirraide commented:
> Looks generally good from a cursory review. I think we want to add tests for modules / in the test/PCH directory.

We already have tests for PCHs, but I can add some more for modules.

> We probably want that as an extension indeed, no reason not to

We allow `friend` type declarations in all language modes, so I’ll do that too.

> We probably want a separate PR to add the pack indexing macro. I did not add it initially because i was concerned the implementation was not mature enough, but i think we fond most of the bugs.
> 
> I think it's fine to set the feature macro for that now with the assumption that we will find all the bugs in the next 6 months.

I’ll open an NFC pr for that.

> There seem to be some missing components, like the json node printer.

I always forget about something when I add an AST node...

Sirraide committed:
[NFC] auto -> auto*

Sirraide committed:
[Clang] Add feature test macro and compat warnings

Sirraide committed:
Merge branch 'main' into variadic-friends

Sirraide committed:
Handle more invalid cases

Sirraide committed:
clang-format

Sirraide commented:
So after talking over some of the more horrible edge cases w/ Corentin for a while, it looks like implementing this requires supporting unexpanded packs in nested name specifiers, which we don’t, but he has a patch for that, so I’ll be taking a look at that next.

Sirraide committed:
clang-format, again

cor3ntin commented:
> So after talking over some of the more horrible edge cases w/ Corentin for a while, it looks like implementing this requires supporting unexpanded packs in nested name specifiers, which we don’t, but he has a patch for that, so I’ll be taking a look at that next.

I think it would be fine and probably preferable to do that feature in stages / multiple PRS

1. This
2. Nested dependent non-pack friends
3. Packs

This is big enough to review 
(If we go that way we can remove the changelog/cxx_status/feature test macro for now)

Sirraide commented:
> I think it would be fine and probably preferable to do that feature in stages / multiple PRS

That makes sense to me, yeah. So the next step would be supporting dependent friend declarations properly, I take it? I.e. cases such as
```c++
template <typename T>
friend class Foo<T>::Nested;
```

And yeah, in that case, I’ll remove the release note etc. for now.

Sirraide committed:
Remove feature test macro etc in preparation of splitting the implementation of this into multiple prs

Sirraide committed:
Add test for cwg2917

Sirraide committed:
Run make-cxx-dr-status

Sirraide committed:
Update JSONNodeDumper

Sirraide commented:
Ok, I might be missing something, but I think this part of the implementation is mostly ‘done’ now

cor3ntin commented:
> Is there ever a situation that would cause us to have to instantiate a FriendPackDecl in the TemplateDeclInstantiator? We do that for UsingPackDecls, but those are sufficiently different enough from the former that I’m not sure about it. I at least haven’t been able to come up w/ any cases that would cause us to try and instantiate one.

That would be needed for the 

```cpp
template <typename Ts>
struct S {
  friend class S<Ts>::Nested...;
};
```

case, I think?

Sirraide committed:
Add more tests

Sirraide committed:
So long, FriendPackDecl

Sirraide commented:
It turns out that `FriendPackDecl` wasn’t needed after all, so that already simplifies this pr quite a bit.

cor3ntin commented on clang/lib/Parse/ParseDeclCXX.cpp:
```diff
@@ -3048,6 +3087,66 @@ Parser::DeclGroupPtrTy Parser::ParseCXXClassMemberDeclaration(
+      return false;
+    };
+
+    if (ParsedFriendDecl(DS))
+      return nullptr;
+
+    if (Tok.is(tok::comma))
+      DiagnoseCompat();
+
+    while (TryConsumeToken(tok::comma)) {
+      ParsingDeclSpec DeclSpec(*this, TemplateDiags);
+      const char *PrevSpec = nullptr;
+      unsigned DiagId = 0;
+      DeclSpec.SetFriendSpec(FriendLoc, PrevSpec, DiagId);
+      ParseDeclarationSpecifiers(DeclSpec, TemplateInfo, AS,
+                                 DeclSpecContext::DSC_class, nullptr);
+      if (ParsedFriendDecl(DeclSpec))
+        return nullptr;
+    }
```
do {} while here?

Sirraide replied:
I’ve cleaned up the code around this a bit, but I’m not sure I can restructure this to a do while loop because the argument to `ParsedFriendDecl` in the loop is different (and I don’t think you can reuse a `ParsingDeclSpec`, or can you?)

cor3ntin replied:
Oh well, good enough!

cor3ntin commented on clang/lib/Sema/SemaDeclCXX.cpp:
```diff
@@ -17354,12 +17354,18 @@ Decl *Sema::BuildStaticAssertDeclaration(SourceLocation StaticAssertLoc,
   bool IsMemberSpecialization = false;
   bool Invalid = false;
 
+  // FIXME: This works for now; revisit once we support packs in NNSs.
```
We can remove this comment

cor3ntin commented on clang/lib/Sema/SemaDeclCXX.cpp:
```diff
@@ -17354,12 +17354,18 @@ Decl *Sema::BuildStaticAssertDeclaration(SourceLocation StaticAssertLoc,
   bool Invalid = false;
 
+  // FIXME: This works for now; revisit once we support packs in NNSs.
+  if (EllipsisLoc.isValid())
+    Diag(EllipsisLoc, diag::err_pack_expansion_without_parameter_packs)
+        << SourceRange(FriendLoc, NameLoc);
+
```
Wait, which case is this?

Sirraide replied:
Iirc, ‘all of the invalid ones involving templates’ for want of a better way of putting it. This seems to work for any friend declarations w/ a template-head that we can’t handle anyway (because the NNS contains an unexpanded pack or is dependent). This is not a particularly good way of doing it (hence the FIXME) and it’s basically where I left off when trying to figure out all the possible code paths where we need to issue an error about a variadic friend declaration being invalid.

I’ll take another look at this once we have clarification on the cwg issue you opened earlier. Also, on that note, how would I check whether a friend declaration is trying to expand a pack that is part of the declaration itself, as opposed to some other pack?

cor3ntin replied:
roughly: collect unexpanded packs. Match them to their declaration and see if that declaration is one if the template parameter of the current declaration

cor3ntin replied:
https://github.com/cplusplus/CWG/issues/593#issuecomment-2271856753 for the proposed wording

Sirraide replied:
Ok, should be done now. That also gets rid of the FIXME and the janky diagnostic here.

Sirraide commented on clang/lib/AST/DeclPrinter.cpp:
```diff
@@ -487,7 +487,26 @@ void DeclPrinter::VisitDeclContext(DeclContext *DC, bool Indent) {
 
     this->Indent();
-    Visit(*D);
+
+    // Group friend declarations if need be.
+    if (isa<FriendDecl>(*D)) {
+      auto *FD = cast<FriendDecl>(*D);
+      VisitFriendDecl(FD);
+      SourceLocation FriendLoc = FD->getFriendLoc();
+
+      // Use a separate iterator; 'D' is always one declaration 'behind' in
+      // this loop; the last friend printed here (or the first printed just
+      // now before this loop if there are no subsequent friends) will be
+      // skipped by the '++D' of the outer loop.
+      for (DeclContext::decl_iterator It; It = std::next(D), It != DEnd; ++D) {
+        auto *NextFriend = dyn_cast<FriendDecl>(*It);
+        if (!NextFriend || NextFriend->getFriendLoc() != FriendLoc)
+          break;
+        VisitFriendDecl(NextFriend, false);
+      }
+    } else {
+      Visit(*D);
+    }
```
Hmm, the problem is that in the AST, we just have multiple `FriendDecl`s in a row with no way of getting from one decl to the next one (that I know of, maybe there’s something I don’t know about, but even then, this function would have to know about it and just skip the remaining friend decls or sth like that...), so we’d have to pass the decl iterator to `VisitFriendDecl`, which feels a bit weird to me. 

Alternatively, we could also move this into a separate helper.

cor3ntin replied:
Oh, I see. I misread that
We don't seem to try to be clever for using declarations, i think we should not here either.

Sirraide replied:
Alright, that works for me :+1:

Sirraide committed:
Address most of Corentin’s feedback

Sirraide committed:
Handle invalid pack expansions properly

Sirraide committed:
clang-format

Sirraide committed:
Update JSON/TextNodeDumper

Sirraide committed:
Revert "Remove feature test macro etc in preparation of splitting the implementation of this into multiple prs"

This reverts commit bf437ab1cfaf5a690a91b7d06325eec2e04f0842.

Sirraide committed:
Remove duplicate feature-test macro

Sirraide committed:
This is already tested elsewhere

Sirraide commented:
Ok, I think I’ve addressed everything now. I’ve also added the feature-test macro back since we won’t be splitting this into multiple prs after all (there are still friend declarations that we don’t handle properly, but that encompasses all friend declarations w/ a dependent NNS irrespective of whether they’re pack expansions or not, so that’s a separate problem).

cor3ntin commented on clang/lib/Sema/SemaDeclCXX.cpp:
```diff
@@ -17556,11 +17584,12 @@ Decl *Sema::ActOnFriendTypeDecl(Scope *S, const DeclSpec &DS,
 
   Decl *D;
   if (!TempParams.empty())
+    // TODO: Support variadic friend template decls?
```
Can we create an issue for that?

Sirraide replied:
I can do that yeah. We just generally don’t support this type of friend declaration, irrespective of whether it’s a pack expansion or not. I can open an issue for that in general if there isn’t already one, because once that’s implemented, supporting pack expansions should be straight-forward.

Sirraide replied:
I’ve opened an issue for supporting friend declarations w/ dependent NNSs: #104057

Once we actually start using `FriendTemplateDecl`s properly (which from what I can tell is what’s required for that), then this should be a natural follow-up.

cor3ntin commented on clang/lib/Sema/SemaDeclCXX.cpp:
```diff
@@ -17509,8 +17528,17 @@ Decl *Sema::ActOnFriendTypeDecl(Scope *S, const DeclSpec &DS,
+      !TSI->getType()->containsUnexpandedParameterPack()) {
+    Diag(EllipsisLoc, diag::err_pack_expansion_without_parameter_packs)
+        << TSI->getTypeLoc().getSourceRange();
+    EllipsisLoc = SourceLocation();
```
Should we mark D as an invalidDecl instead?

Sirraide replied:
I mean, we weren’t doing that before either, but I can try doing that.

cor3ntin replied:
Thanks

Sirraide replied:
Done. That seems to have worked.

cor3ntin reviewed (commented):
Can you make a separate pr to update the unrelated changes to cxx_status?
I think we are pretty close to be able to land this

Sirraide commented:
> Can you make a separate pr to update the unrelated changes to cxx_status?

Ah, yeah, I didn’t notice those. I’ll get those removed from this pr.

Sirraide committed:
Merge branch 'main' into variadic-friends

Sirraide committed:
Mark decl as invalid

Sirraide committed:
Remove unrelated changes from cxx_dr_status

Sirraide commented:
> I’ll get those removed from this pr.

Also done. I think that’s everything now?

cor3ntin reviewed (approved):
LGTM

Endilll commented:
@Sirraide CWG looked at 2917 last week, and changed the status of the issue. Can you revisit your test for it? 
https://cplusplus.github.io/CWG/issues/2917.html

Sirraide commented:
> CWG looked at 2917 last week, and changed the status of the issue. Can you revisit your test for it?

Ah, something that was previously accepted is now disallowed; I’ll implement that tomorrow.

cor3ntin commented:
I would recommend creating an issue and see how ewg handles the new information before reaping out a lot of code.

Endilll commented:
Note that there have been additional discussions happening on the new proposed resolution.
https://lists.isocpp.org/core/2024/08/16236.php

Sirraide commented:
> I would recommend creating an issue and see how ewg handles the new information before reaping out a lot of code.

Hmm, I mean, it looks like they’ve already thought about the consequences though, or am I missing something here, because the case that we currently support and which this ends up disallowing is explicitly listed as disallowed in the core issue:
> The proposed resolution above disallows a few examples from paper [P2893R3](https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2024/p2893r3.html) (Variadic friends), such as
> ```c++
>  template<class... Ts>
>  struct VS {
>    template<class U>
>    friend class C<Ts>::Nested...; // now ill-formed
>  };
> ```

> https://lists.isocpp.org/core/2024/08/16236.php

Ah, that I don’t appear to have access to unfortunately.

Endilll commented:
> I would recommend creating an issue and see how ewg handles the new information before reaping out a lot of code.

Sure, we don't necessarily need to change our implementation as CWG continues to consider the issue, but we need to update the test and the status for bookkeeping purposes. That's only fair when we write tests against moving targets (open issues).

Sirraide commented:
> we need to update the test and the status for bookkeeping purposes

In that case, I’ll leave the implementation of this as is for now and just add a FIXME and update the dr status.

cor3ntin commented:
I think the dr test already say "Clang 20 implements 2024-07-30 resolution" which is correct

Endilll commented:
> I think the dr test already say "Clang 20 implements 2024-07-30 resolution" which is correct

While I understand what you're referring to (author's claim that the example is nonsense and should be rejected), if you open https://cplusplus.github.io/CWG/issues/2917.html, the only resolution there is dated 2024-08-16.

If for whatever reason we think that we should stick to that instead of considering newer proposed resolution, I'm happy with a simple replacement of `open` with `review` in the status.

cor3ntin commented:
the date changed. weird! the resolution did not

On Fri, Aug 23, 2024, 20:59 Vlad Serebrennikov ***@***.***>
wrote:

> I think the dr test already say "Clang 20 implements 2024-07-30
> resolution" which is correct
>
> While I understand what you're referring to (author's claim that the
> example is nonsense and should be rejected), if you open
> https://cplusplus.github.io/CWG/issues/2917.html, the only resolution
> there is dated 2024-08-16.
>
> If for whatever reason we think that we should stick to that instead of
> considering newer proposed resolution, I'm happy with a simple replacement
> of open with review in the status.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/llvm/llvm-project/pull/101448#issuecomment-2307646855>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAKX7632RFGI2VOAS4H3WK3ZS6A77AVCNFSM6AAAAABLZVR77OVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDGMBXGY2DMOBVGU>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

Sirraide commented:
Pretty sure it did yeah; the change is that this case is now no longer valid: https://github.com/llvm/llvm-project/pull/101448#issuecomment-2304770725

Endilll commented:
Jens restored the suggested resolution in https://cplusplus.github.io/CWG/issues/2917.html and even put a date on it in case we need to reference it.

I also notified him that in our tests we are referring to proposed and suggested resolution by dates.
He also mentioned we can't rely on git history of https://github.com/cplusplus/CWG, because he's regularly squashing it. So if we think that something is wrong with CWG issue pages, we should bring this up to him instead of trying to cope using git history.

cor3ntin commented:
we are implementing the proposed resolution, which is the second set of change in that issue. we should *not* implement the "suggested resolution" as discussions are still ongoing (notably concerns of it being too restrictive).

Just updating the date is fine

Sirraide commented:
> we are implementing the proposed resolution, which is the second set of change in that issue. we should _not_ implement the "suggested resolution" as discussions are still ongoing (notably concerns of it being too restrictive).

Well, the suggested resolution, i.e. *accepting*
```c++
template<class... Ts>
struct VS {
    template<class U>
    friend class C<Ts>::Nested...; // now ill-formed
};
 ```

is what’s already implemented. The proposed resolution rejects this instead.
